### PR TITLE
UnitSAO: Prevent circular attachments

### LINKED ISSUE #11079: Crash when trying to attach parent object to its child
##### Minetest version
```
5.4.0
```

##### Summary
If you try to attach a parent object (an object to which another object is attached to) to its child, Minetest first freezes, than crashes completely with a segfault.

##### Steps to reproduce
1. Open DevTest
2. Place a `testentities:cube` and a `testentities:armorball` with the entity spawner
3. Pick up the object attacher
4. Attach the cube to the armor ball (first punch the cube, then the armor ball) → the armor ball should now be the parent of the cube
5. Verify you did it correctly by punching the armor ball with the children getter tool
6. Now try to attach the armor ball to the cube (first punch the cube with the object attacher, then the armor ball)

##### Note
First, the armor ball is the parent of the cube, but in the next step it is tried to make also the cube a parent of the child. Of course, this is a logical mistake, as this creates a reference loop that can never be resolved. Good mods should never make this kind of error, but since Minetest segfaults here, this is a Minetest bug.

The bug can also be triggered if the the reference loop is more complicated with multiple layers of parents/children (e.g. A → B → C → D → E → A)

### PR DESCRIPTION
**Pull request by SmallJoker**
Fixes #11079

## To do

This PR is Ready for Review.

## How to test

> 1. Open DevTest
>2. Place a `testentities:cube` and a `testentities:armorball` with the entity spawner
>3. Pick up the object attacher
>4. Attach the cube to the armor ball (first punch the cube, then the armor ball) → the armor ball should now be the parent of the cube
>5. Verify you did it correctly by punching the armor ball with the children getter tool
>6. Now try to attach the armor ball to the cube (first punch the ~~cube with the object attacher, then the~~ armor ball, *then the cube*)

Shameless copy from #11079 with corrections

Step 6 should log "Attachment failed!" to the chat.

### CODE REVIEW

**Comment by Wuzzy2** [comment]
I tested this and it works, including with multiple levers of attachment.
But maybe there should be a note in `lua_api.txt` now that `set_attachment` can silently fail (without crash). Otherwise modders might assume that `set_attach` always works and don't check.

File: src/server/unit_sao.cpp [hunk 1]
```diff
@@ -124,6 +124,16 @@ void UnitSAO::sendOutdatedData()
 void UnitSAO::setAttachment(int parent_id, const std::string &bone, v3f position,
 		v3f rotation, bool force_visible)
 {
+	auto *obj = parent_id ? m_env->getActiveObject(parent_id) : nullptr;
```
Comment (nerzhul) [inline src/server/unit_sao.cpp:127]: if (auto *obj = (parent_id ? m_env->getActiveObject(parent_id) : nullptr)) can be better :)

File: src/server/unit_sao.cpp [hunk 2]
```diff
@@ -124,6 +124,16 @@ void UnitSAO::sendOutdatedData()
...
+	if (obj) {
+		// Do checks to avoid circular references
+		// The chain of wanted parent must not refer or contain "this"
+		for (obj = obj->getParent(); obj; obj = obj->getParent()) {
```
Comment (nerzhul) [inline src/server/unit_sao.cpp:131]: we should log if someone try to attach a parent to its child, it's clearly a bug or a modding error which should not be silent

File: src/server/unit_sao.cpp (hunk 1 above)
Comment (SmallJoker) [reply src/server/unit_sao.cpp:127]: Same assembly code, same speed, but less readable IMO. Thanks for the suggestion, but I'd prefer to stick to this style.

File: src/server/unit_sao.cpp (hunk 2 above)
Comment (SmallJoker) [reply src/server/unit_sao.cpp:131]: warningstream added

File: doc/lua_api.txt [hunk 3]
```diff
@@ -6314,6 +6314,8 @@ object you are working with still exists.
       Default `{x=0, y=0, z=0}`
     * `forced_visible`: Boolean to control whether the attached entity
        should appear in first person. Default `false`.
+    * This command may fail silently (do nothing) when it would result
```
Comment (nerzhul) [inline doc/lua_api.txt:6317]: this comment is invalid now no ? as you added a warning stream

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: doc/lua_api.txt (hunk 3 above)
Comment (SmallJoker) [reply doc/lua_api.txt:6317]: There's no return value or error to indicate the failure, hence it's a silent fail in perspective of the modding API.

**Pull request merged**